synapse/rest/admin/rooms.py

```diff
@@ -464,6 +464,10 @@ async def on_POST(
         assert_params_in_dict(content, ["user_id"])
         target_user = UserID.from_string(content["user_id"])
 
+        historical = False
+        if "historical" in content:
+            historical = content["historical"]
```

Comment:
Unfortunately I don't think this really makes sense for mainline. While setting `historical` to True here would prevent push notifications from being generated by this event, it would also prevent this event from coming down `/sync`, which is generally not what you want when you join a user to a room.

The `is_historical` boolean originates from MSC2716 (an experimental MSC), and arguably https://github.com/matrix-org/synapse/blob/848f7e3d5ff38ca28e56e6143e584974da1eec42/synapse/push/bulk_push_rule_evaluator.py#L334 should also be gated behind the `experimental.msc2716_enabled` config option.

I think this *would* work for your use case, as the join events you intend to generate will be in a room that is hidden from the client anyways (and will probably work fine for your use case). But this change would unfortunately not make sense for mainline Synapse.

Reply:
Maybe we can set historical= false in case the room is in exclude_rooms_from_sync

Reply:
For my understanding `/sync` is to synchronize the state of room through events.

1/ Therefore if we set `historical` to True, the existing users won't even know that a new joiner has come into the room and Clients(mobile and web) will not be updated unless if they do a full refresh, is this correct?

2/ Another question is in `_local_membership_update` called by `self.room_member_handler.update_membership` allows to past an historical event : in which case do we update membership as `historical` event?
As a result, are we sure that we do not want a `/join` to be considered as an `historical` event (for Admin API) regardless of the `/sync` as we do not want to notify the state to clients ?

3/ Perhaps, it would be a good idea to add a new parameter to do not notify users when doing the `/join`, could it be a new feature? There is a `is_notifiable` on an event to help.

Reply:
> 1/ Therefore if we set `historical` to True, the existing users won't even know that a new joiner has come into the room and C

That is correct, yes, unless they do a `/messages` request for historical messages to find out.

> 2/ Another question is in `_local_membership_update` called by `self.room_member_handler.update_membership` allows to past an historical event : in which case do we update membership as `historical` event? 

Yes.

> As a result, are we sure that we do not want a `/join` to be considered as an `historical` event (for Admin API) regardless of the `/sync` as we do not want to notify the state to clients ?

Typically you do want clients to know that a user has joined a room though, so that the new user appears in the room list, can be mentioned, can send message keys via to-device messages etc.

Whether clients should get a notification is a different story, but a client *may* set up their push rules so that they except to receive a push notification when a user joins a room.

> 3/ Perhaps, it would be a good idea to add a new parameter to do not notify users when doing the `/join`, could it be a new feature? There is a `is_notifiable` on an event to help on the latest codebase.

Yes, I think a more generic option would be the way to go. I don't think we should abuse `historical` for this purpose (it may be removed from Synapse in a later update anyways). I suggest making use of it for this purpose but then reverting the change afterwards after all existing users have been added to the discovery rooms.

Reply:
We have done some testing and we have find out that this patch resolves the performance issue but it does not fix the research of "federated" users.
We will give up this solution in favor of : #2 which in terms of performance might take a bit longer but still better than doing nothing.
Also, if we do not switch off the push rules we will not be able to make a user /join on a daily process, so the patch has to remain in the code.